src/main/java/no/digipost/io/InputStreamIterator.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) Posten Norge AS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package no.digipost.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * InputStreamIterator is an {@link Iterator} reading from an {@link InputStream} in chunks
+ * where each chunk is returned as the next element in the iterable.
+ * When the input stream is fully consumed the iterator has no more elements.
+ */
+public class InputStreamIterator implements Iterator<byte[]>, AutoCloseable {
+    private final InputStream inputStream;
+    private final int chunkSize;
+    private byte[] next;
+    private Boolean hasNext;
+    private boolean endOfStreamReached = false;
+
+    public InputStreamIterator(InputStream inputStream, int chunkSize) {
+        this.inputStream = inputStream;
+        this.chunkSize = chunkSize;
+    }
+
+    private byte[] loadNextChunk() {
+        if (endOfStreamReached) return null;
+
+        byte[] chunk = new byte[chunkSize];
+        int bytesRead = 0;
+        try {
+            bytesRead = inputStream.read(chunk);
+            if (bytesRead == -1) {
+                endOfStreamReached = true;
+                inputStream.close();
+                return null;
+            }
+        } catch (IOException e) {
+            throw new WrappedInputStreamFailed(e, inputStream);
```

Comment:
Here, I think I prefer to wrap the caught IOException in an [UncheckedIOException](https://docs.oracle.com/javase/8/docs/api/java/io/UncheckedIOException.html) instead of introducing a new exception type. The reason for this is:
- This location is where you _actually_ know that _reading_ (or even closing) from the wrapped stream failed, but the semantics expressed by the type of the custom exception is just a "general failure in the wrapped stream"
- Still, _inside_ the constructor of `WrappedInputStreamFailed`, it is assumed that this exception will only be instantiated in the case of a reading error on a InputStream. So even though the _type name_ communicates a (reusable) exception type for any failure of wrapped InputStreams, internally it assumes it is only being used in the case of failing to _read_ an InputStream.

Therefore, I think the exception message may as well be constructed in this location, because this is where the knowledge about what failed actually is. And then you may as well reuse an exception applicable exception type, which in this case I think is an UncheckedIOException.

Alternatively, you could express the semantics in the type name of the custom exception (`WrappedInputStreamReadingFailed`), but that would just make it even more evident that the custom exception is really crafted for this single situation, and you may as well just use a more general exception. Unless you are really interested in capturing that specifically _reading_ the wrapped stream failed, not just any IO failure while processing the wrapped stream (which is reading for the most part).